File: LeetCode c#/3-LongestSubstringWithoutRepeatingCharacters.cs
//3. Longest Substring Without Repeating Characters

public class Solution {
    public int LengthOfLongestSubstring(string s) {

        if (s == null || s == String.Empty)
            return 0;

        HashSet<char> set = new HashSet<char>();
        int currentMax = 0;
        int i = 0;
        int j = 0;

        while (j < s.Length)
            if (!set.Contains(s[j]))  {
                set.Add(s[j++]);
                currentMax = Math.Max(currentMax, j - i);
            }
            else
                set.Remove(s[i++]);

        return currentMax;
    }
    
    
}

File: LeetCode Java/Algorithm I/3-LongestSubstringWithoutRepeatingCharacters.java
/*
 * 3. Longest Substring Without Repeating Characters
 * Given a string s, find the length of the longest substring without repeating
 * characters.
 * 
 * Example 1:
    * Input: s = "abcabcbb"
    * Output: 3
    * Explanation: The answer is "abc", with the length of 3.
 */
class Solution {
    // O(n) space and time complexity
    public int lengthOfLongestSubstring(String s) {
        // sliding window
        HashSet<Character> unq = new HashSet<Character>();

        int largest = 0; // output
        int i = 0; // leftside slider
        int j = 0; // right side slider

        // iterate through each letter
        while (j < s.length()) {
            if (!unq.contains(s.charAt(j))) { // add unique characters
                unq.add(s.charAt(j++)); // move right side of the slider
                largest = Math.max(largest, j - i); // take the max seqeuence

            } else {
                unq.remove(s.charAt(i++)); // remove first left side, and move slider
            }

        }

        return largest;
    }
}